plugins/nominal-connection-checker/src/generic_map.js

```diff
@@ -0,0 +1,237 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Defines the GenericMap.
+ */
+'use strict';
+
+import {PriorityQueueMap} from './priority_queue_map';
+
+/**
+ * The priority for binding an explicit type to a generic type based on an input
+ * of the block with the generic type.
+ * @type {number}
+ */
+export const INPUT_PRIORITY = 100;
+
+/**
+ * The priority for binding an explicit type to a generic type based on the
+ * output of the block with the generic type.
+ * @type {number}
+ */
+export const OUTPUT_PRIORITY = 200;
+
+/**
+ * Defines a map where generic types can be bound to explicit types in
+ * a given environment.
+ */
+export class GenericMap {
+  /**
+   * Constructs the GenericMap.
+   * @param {!Blockly.Workspace} workspace The workspace this GenericMap belongs
+   *     to.
+   */
+  constructor(workspace) {
+    /**
+     * The workspace this GenericMap belongs to.
+     * @type {!Blockly.Workspace}
+     * @private
+     */
+    this.workspace_ = workspace;
+
+    /**
+     * Map of block ids to PriorityQueMaps that define what the generic types
+     * of the block are bound to.
+     * @type {!Map<string, !PriorityQueueMap>}
+     * @private
+     */
+    this.dependenciesMap_ = new Map();
+
+    /**
+     * Map of block ids to objects that map generic type names to arrays of
+     * DependerInfo specifying what other blocks and generic types on those
+     * blocks depend on the generic type of the lookup block.
+     * @type {!Map<string, !Map<string, !Array<DependerInfo>>>}
+     * @private
+     */
+    this.dependersMap_ = new Map();
+  }
+
+  /**
+   * Returns the name of the explicit type bound to the generic type in the
+   * context of the given block, or undefined if the type is not bound.
+   * @param {string} blockId The block id that the generic type is
+   *     possibly bound in.
+   * @param {string} genericType The generic type to find the explicit binding
+   *     of.
+   * @return {undefined|string} The name of the explicit type bound to the
+   *     generic type in the context of the given block, or undefined if the
+   *     type is not bound.
+   */
+  getExplicitType(blockId, genericType) {
+    const priorityMap = this.dependenciesMap_.get(blockId);
+    if (!priorityMap) {
+      return undefined;
+    }
+    const types = priorityMap.getValues(genericType);
```

Comment:
Is types guaranteed to have at least one entry if it's not null?

Reply:
I checked, and the PriorityQueueMap handles that [here](https://github.com/google/blockly-samples/pull/376/files/7d8071b16dd6ee5f8713801a4e4fb84c8d10d085). I'm not sure if I handled it correctly for the dependersMap_ though so I'll have to check. And I'll add some tests as well!